### PR TITLE
Number improvements

### PR DESCRIPTION
**Pull request by dario-loi**
# Changes

## Floating Point Numbers!

Added the facility to generate floats in various ways, mirroring the previous capabilities available through `integer`.

## Changes to Number Template

The Number class is not a template anymore, rather, its individual methods are, this should reduce the amount of code generated, as only the individual calls are compiled.

## Custom Distributions

Library consumers can now specify their own distributions to be used by the library to generate numbers. 

Sampling with a custom distribution can be done both by specifying a range, in which case the output will be clamped (the hard clamping is kind of undesirable, since it ruins the statistical properties of the distribution, a continuous resampling until the value falls in the interval might be more desirable, but might lead to very long executions in case most of the distribution's density is not located inside the user specified bounds), and by specifying no range, in which case the sampling will be uncostrained.

Sampling without bounds might look like a useless wrapper, in my opinion it is still a convenient interface since we re-use the random device that we initialized in our class, hence saving our user quite a bit of code that he would need in order to initialize those objects himself.

## More Concept Type Safety

By enforcing a large part of the requirements that a RandomNumberDistribution must have according to the C++ standard, we ensure that while our users can use any distribution to sample through our utilities, they are less likely to design bad code with it, as the new concepts that are introduced would catch their errors at compile time.

## New tests

A bunch of new tests have been created to provide more code coverage for the Number module, including floating point counterparts for the integral tests, and a small number of tests for the new distributions. After the CI/CD pipeline runs and produces a code report, we can understand if we need more.

## After this update

I do not know what I'll do next after this pull request, if I have more time I might design a system to generate the data in src/data from files using CMake, this would allow easier expansion of the databases.

A few suggestions for other small improvements:

* `std::string_view` and `std::array` could be used to allow the stored data to live in static storage, potentially increasing cache friendliness(?) and reducing allocations(?) (should be benchmarked, however, not using `std::string` will surely reduce the number of allocated memory).
* Clang tidy could be integrated in the CI/CD pipeline, together with some modernize FIXITs, so that we can instantly get some refactorings such as trailing return types without much effort. Code quality will also increase as we are going to increase the amount of warnings that can be checked.

### CODE REVIEW

**Comment by dario-loi** [comment]
# Bugfixes & Refactors

there previously was a malicious cast in the Helper.h class, instead of fixing just that, I refactored it to use the modern `std::span` interface, this allows our library to sample elements not just from `std::vector`s, but from stacks, lists, C style arrays, and more!

There is a disadvantage, since template deduction can't both construct a span and deduce the element type at the same time, the user is now required to specify what type is contained into the container that is being passed to `ArrayElement`, I find it quite redundant and a bit verbose, so I am going to point this out. 

In conclusion the change massively increases the flexibility of the library, but introduces a bit more verbosity in the mix.

File: include/faker-cxx/Number.h [hunk 1]
```diff
@@ -1,55 +1,378 @@
...
+ *
+ * @date 2023-07-05
+ *
+ *  @details The module provides functions for both integer and decimal numbers, each with a number of overloads.
+ *  There are generally functions that allow the generation of uniformly distributed numbers within a range,
+ *  and functions that allow the generation of numbers from a given distribution, both in a given range and not.
+ * 
+ *  There is a possibility to only specify the right bound of the range, in which case the left bound is assumed to be 0.
+ */
+
```
Comment (cieslarmichal) [inline include/faker-cxx/Number.h:14]: What do you think about documenting only things that users use? I mean not to document each file, but document classes and classes methods.

File: src/Number.cpp [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#include "Number.h"
+
+namespace faker
+{
+
+/**
+ * @brief A random device used to seed the Number module pseudo-random generator.
```
Comment (cieslarmichal) [inline src/Number.cpp:7]: I dont think we need those docs about randomDevice and pseudoRandomGenerator, we need to be careful to write only essential comments/docs because it would be hard to modify and read later if we make many of them

**Review by cieslarmichal — COMMENTED** [review]
A bit hardcore changes, nice one :) I think there are too many docs (file docs, member docs), what do you think about this?

File: src/Number.cpp (hunk 2 above)
Comment (dario-loi) [reply src/Number.cpp:7]: You are right, it is best to document only the essential things, otherwise the important stuff gets lost in the noise, you can go ahead and clean this

File: include/faker-cxx/Number.h (hunk 1 above)
Comment (dario-loi) [reply include/faker-cxx/Number.h:14]: It might be okay, but documenting the files allows the individual headers to have a doxygen documentation, so if one opens up the html docs he can get an overview of what each module does.

If you feel like it is too much clutter you can go ahead and remove the file docs, but I think they are pretty standard in C++
